### PR TITLE
feat(datastructure): update

### PR DESCRIPTION
**Pull request by acateland**
Change flow datastructure.
This change of flow datastructure gear this lib toward a more specific use inside talend and less as a general purpose library

All visual component (Components, Ports, Links) see theyr attributes property set into two :
- ```data``` wich target to hold functional data needed by the backend
- ```graphicalAttribtues``` who are solely usefull for the frontend

data and graphicalAttributes see the creation of a properties bag wich is intended to store any data you want

Node see its following attributes shuffled around:
- NodeSize moved from the root to graphicalAttributes
- position moved from the root to graphicalAttributes
- nodeType used to resolve witch component to use for rendering to graphicalAttributes
- type wich is more general value set to the root of the component
- label is moved to graphicalAttributes 

Port see its following attributes shuffled around:
- position is moved inside graphicalAttributes
- portType is moved inside graphicalAttributes

Port see its following attributes shuffled around:
- linkType is moved inside graphicalAttributes

### CODE REVIEW

File: src/components/ZoomHandler.component.jsx [hunk 1]
```diff
@@ -30,10 +30,14 @@ const ZoomHandler = React.createClass({
 		}
 	},
 	onZoomEnd() {
-		this.props.setZoom(event.transform);
+		if (!isNaN(event.transform.k) && !isNaN(event.transform.x) && !isNaN(event.transform.y)) {
```
Comment (jmfrancois) [inline src/components/ZoomHandler.component.jsx]: event here is global (from import)
I don't understand the need to check if k, x & y are numbers.
It looks like pre assertion, should be the setZoom which is doing this check;

File: src/components/ZoomHandler.component.jsx [hunk 2]
```diff
@@ -30,10 +30,14 @@ const ZoomHandler = React.createClass({
...
+			this.props.setZoom(event.transform);
+		}
 	},
 	onZoom() {
-		this.setState({ transform: event.transform });
+		if (!isNaN(event.transform.k) && !isNaN(event.transform.x) && !isNaN(event.transform.y)) {
```
Comment (jmfrancois) [inline src/components/ZoomHandler.component.jsx]: same here

File: src/components/link/AbstractLink.component.jsx [hunk 3]
```diff
@@ -45,41 +45,41 @@ const AbstractLink = React.createClass({
 			nextProps.target !== this.props.target ||
 			nextProps.targetHandlePosition !== this.props.targetHandlePosition;
 	},
-	renderLinkSourcehandle(){
-		if(this.props.linkSourceHandleComponent){
-			return <LinkHandle
+	renderLinkSourcehandle() {
```
Comment (jmfrancois) [inline src/components/link/AbstractLink.component.jsx]: I think this could be a pure component outside of this one which just group + render LinkSourceHandler and LinkTargetHandler

By the way did you do a typo here ? handle**R** ?

File: src/components/link/AbstractLink.component.jsx [hunk 4]
```diff
@@ -45,41 +45,41 @@ const AbstractLink = React.createClass({
...
+		if (this.props.linkSourceHandleComponent) {
+			return (<LinkHandle
 				component={this.props.linkSourceHandleComponent}
 				onDrag={this.props.onSourceDrag} onDragEnd={this.props.onSourceDragEnd}
-				position={this.props.sourceHandlePosition || this.props.source.position}
-			/>
+				position={this.props.sourceHandlePosition || this.props.source.getIn(['graphicalAttributes', 'position'])}
```
Comment (jmfrancois) [inline src/components/link/AbstractLink.component.jsx]: this is quite hard to read. May be some getter would be helpful here.

File: src/components/node/AbstractNode.component.jsx [hunk 5]
```diff
@@ -18,110 +18,113 @@ import { PositionRecord } from '../../constants/flowdesigner.model';
...
+	render() {
+		const { node } = this.props;
+		const x = node.getIn(['graphicalAttributes', 'position', 'x']);
+		const y = node.getIn(['graphicalAttributes', 'position', 'y']);
+		const transform = `translate(${x}, ${y})`;
+		return (
+			<g>
+				<g
+					transform={transform}
+					ref={c => (this.nodeElement = c)} onClick={this.onClick}
```
Comment (jmfrancois) [inline src/components/node/AbstractNode.component.jsx]: style: onClick should go the line after

File: src/reducers/flow.reducer.js [hunk 6]
```diff
@@ -58,15 +58,15 @@ export const reducer = (state, action) => {
 			);
 		} catch (error) {
 			invariant(
-				true,
+				false,
```
Comment (jmfrancois) [inline src/reducers/flow.reducer.js]: WAT ?

File: src/reducers/flow.reducer.js [hunk 7]
```diff
@@ -103,12 +103,13 @@ export const calculatePortsPosition = (state, action) => {
...
 		return nodes.reduce((cumulativeState, node) => {
+			const nodeType = node.getIn(['graphicalAttributes', 'nodeType']);
 			const ports = state.get('ports').filter(port => port.nodeId === node.id);
-			const calculatePortPosition = state.getIn(['nodeTypes', node.nodeType, 'component'])
+			const calculatePortPosition = state.getIn(['nodeTypes', nodeType, 'component'])
 				.calculatePortPosition;
 			return cumulativeState.mergeIn(
 				['ports'],
-				calculatePortPosition(ports, node.position, node.nodeSize),
+				calculatePortPosition(ports, node.getIn(['graphicalAttributes', 'position']), node.getIn(['graphicalAttributes', 'nodeSize'])),
```
Comment (jmfrancois) [inline src/reducers/flow.reducer.js]: style: line is too long here (129)

File: src/reducers/node.reducer.test.js [hunk 8]
```diff
@@ -7,33 +7,44 @@ import { NodeRecord, PositionRecord } from '../constants/flowdesigner.model';
...
+		graphicalAttributes: new Map({
+			type: 'type1',
+			selected: true,
+			position: new PositionRecord({ x: 10, y: 10 }),
+		}),
 	})).setIn(['nodes', 'id2'], new NodeRecord({
 		id: 'id2',
-		nodeType: 'type2',
+		type: 'type2',
 		position: new PositionRecord({ x: 10, y: 10 }),
```
Comment (jmfrancois) [inline src/reducers/node.reducer.test.js]: position should not be here (it's in graphicalattributes now)

File: src/reducers/port.reducer.test.js [hunk 9]
```diff
@@ -10,15 +10,15 @@ describe('Check port reducer', () => {
 			id: 'id1',
 			position: new PositionRecord({ x: 10, y: 10 }),
```
Comment (jmfrancois) [inline src/reducers/port.reducer.test.js]: should not be exists here

File: src/reducers/port.reducer.test.js [hunk 10]
```diff
@@ -27,61 +27,57 @@ describe('Check port reducer', () => {
...
-	it('FLOWDESIGNER_PORT_ADDS to add multiple ports to port collection', () => {
+	it('FLOWDESIGNER_PORT_ADDS to add multiple ports (portId1, portId2) to port collection', () => {
 		expect(portReducer(initialState, {
 			type: 'FLOWDESIGNER_PORT_ADDS',
 			nodeId: 'nodeId',
 			ports: [{
 				portId: 'portId1',
-				portType: 'portType',
-				data: { type: 'EMITTER' },
+				graphicalAttributes: { portType: 'portType', properties: { type: 'EMITTER' } },
```
Comment (jmfrancois) [inline src/reducers/port.reducer.test.js]: question: in reducer you have javascript object structure  for graphicalAttributes ?

Comment (acateland) [inline src/reducers/port.reducer.test.js]: this is a mocked action where datastructure are kept in simple literral javascript object

File: src/components/link/AbstractLink.component.jsx (hunk 3 above)
Comment (acateland) [inline src/components/link/AbstractLink.component.jsx]: no handle as a door handler

File: src/components/ZoomHandler.component.jsx (hunk 1 above)
Comment (acateland) [inline src/components/ZoomHandler.component.jsx]: did this because it happen that on initialisation event is returnerd with NaN value set on x/y, guess this is because canvas size is not yet determined by the browser (code initialisation is faster in some case than layout composition)

File: src/components/link/AbstractLink.component.jsx (hunk 4 above)
Comment (acateland) [inline src/components/link/AbstractLink.component.jsx]: im going to do this on another PR https://github.com/Talend/react-flow-designer/issues/192

File: src/reducers/flow.reducer.js (hunk 6 above)
Comment (acateland) [inline src/reducers/flow.reducer.js]: reverted don't remember why i did this, maybe for debug purpose

Comment (acateland) [inline src/reducers/flow.reducer.js]: ok remember why i did this, if new error is thrown it prevent return a default state breaking the app
